src/Microsoft.ML.Data/Transforms/RowToRowTransformerBase.cs

```diff
@@ -23,7 +23,8 @@ protected RowToRowTransformerBase(IHost host)
             Host = host;
         }
 
-        public abstract void Save(ModelSaveContext ctx);
+        void ICanSaveModel.Save(ModelSaveContext ctx) => SaveModel(ctx);
+        private protected abstract void SaveModel(ModelSaveContext ctx);
```

Comment:
>private protected abstract void SaveModel(ModelSaveContext ctx); [](start = 8, length = 64)

Even though we might not be making use of it right now, I think we still ought to make this `[BestFriend]` to signal at least to internal developers that this is *intended* as something that someone implementing this, that this is definitely an intended expansion point.

Reply:
Actually, wait, how is this actually working? I see `ImageGrayscalingTransformer`, which I *think* is overriding this method. That's definitely in an entirely separate assembly. So how is this working without this being declared a `[BestFriend]`? (I'm sort of working via scrolling through the text of your review, so I could easily have missed something, but I have to expect that this is the root of all the basic transformer functionality.) If it is somehow working, which I don't fully understand how it could, this means that there's probably some bug or unintended behavior in `[BestFriend]`? Or am I wrong?

---
In reply to: [254475884](https://github.com/dotnet/machinelearning/pull/2431#discussion_r254475884) [](ancestors = 254475884)

Reply:
I guess *overriding* a method doesn't count as a *reference* to it, so the analyzer rule wouldn't be triggered. Huh. Not sure if I should care or not...

Reply:
It is the core of the transformers, so yes I guess that this is not being picked up by the analyzer.
But I agree in making it BestFriend. 

---
In reply to: [254554561](https://github.com/dotnet/machinelearning/pull/2431#discussion_r254554561) [](ancestors = 254554561)